libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/builder/NavRouteOptionsBuilder.kt

```diff
@@ -0,0 +1,362 @@
+package com.mapbox.navigation.core.routeoptions.builder
+
+import android.content.Context
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.Bearing
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.extensions.applyDefaultNavigationOptions
+import com.mapbox.navigation.base.extensions.applyLanguageAndVoiceUnitOptions
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface NoWaypointsOptionsBuilder {
+    fun fromStartLocation(
+        point: Point,
+        bearing: Double? = null,
+        zLevel: Int?
+    ): WaypointsInProgressBuilder
+
+    fun fromCurrentLocation(): WaypointsInProgressBuilder
+}
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface WaypointsInProgressBuilder {
+    fun addIntermediateWaypoint(
+        coordinate: Point,
+        name: String? = null,
+        bearing: Double? = null,
+        zLevel: Int? = null,
+        targetCoordinate: Point? = null
+    ): WaypointsInProgressBuilder
+
+    fun addIntermediateSilentWaypoint(
+        coordinate: Point,
+        bearing: Double? = null,
+        zLevel: Int? = null,
+    ): WaypointsInProgressBuilder
+
+    fun toDestination(
+        coordinate: Point,
+        name: String? = null,
+        bearing: Double? = null,
+        zLevel: Int? = null,
+        targetCoordinate: Point? = null
+    ): RouteOptionsBuilderWithWaypoints
+}
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface RouteOptionsBuilderWithWaypoints {
+    fun profileDriving(
+        drivingSpecificSetup: DrivingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+
+    fun profileDrivingTraffic(
+        drivingSpecificSetup: DrivingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+
+    fun profileWalking(
+        walkingSpecificSetup: WalkingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+
+    fun profileCycling(
+        cyclingSpecificSetup: CyclingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+}
```

Comment:
I see what you mean, the API would look less messy if we used one of two approaches.  I don't want to get rig of the step builder for now, because it makes the API safe. Anyway let's discuss the step bulder in the [different answer](https://github.com/mapbox/mapbox-navigation-android/pull/5427#discussion_r806673433), maybe we will decide to get rid of it, who knows 🙂 

@tomaszrybakiewicz , would you like the API more if I replaced configuration block by builder?
```kotlin
val routeOptions = mapboxNavigation.buildRouteOptions { builder ->
                builder
                    .fromCurrentLocation()           
                    .toDestination(destination)      
                    .profileDriving { drivingProfileConfig ->                            
                        drivingProfileConfig
                            .maxWidth(1.42)
                            .maxHeight(2.2)             
                    }
            }
```

Reply:
Still looks awkward. But at least it's consistent with `buildRouteOptions`.

Reply:
Thank you for feedback. I understand what you feel and I tend to agree with you. I will think about other way to structure API. 👍 
But I think that current API is okay for the first iteration. As the feature is experimental, I will be able to change API in the future.